farm-staking/farm-staking/src/external_interaction.rs

```diff
@@ -0,0 +1,186 @@
+        let caller = self.blockchain().get_caller();
+        self.require_user_whitelisted(&user, &caller);
+
+        let payments = self.get_non_empty_payments();
+        self.check_additional_payments_original_owner(&user, &payments);
+
+        let boosted_rewards = self.claim_only_boosted_payment(&user);
+        let boosted_rewards_payment =
+            EsdtTokenPayment::new(self.reward_token_id().get(), 0, boosted_rewards);
+
+        let enter_result = self.enter_farm_base::<FarmStakingWrapper<Self>>(user.clone(), payments);
+
+        let new_farm_token = enter_result.new_farm_token.payment.clone();
+        self.send_payment_non_zero(&caller, &new_farm_token);
+        self.send_payment_non_zero(&user, &boosted_rewards_payment);
+
+        self.set_farm_supply_for_current_week(&enter_result.storage_cache.farm_token_supply);
+
+        self.update_energy_and_progress(&user);
+
+        self.emit_enter_farm_event(
+            &caller,
+            enter_result.context.farming_token_payment,
+            enter_result.new_farm_token,
+            enter_result.created_with_merge,
+            enter_result.storage_cache,
+        );
+
+        (new_farm_token, boosted_rewards_payment).into()
+    }
+
+    #[payable("*")]
+    #[endpoint(claimRewardsOnBehalf)]
+    fn claim_rewards_on_behalf(&self) -> ClaimRewardsResultType<Self::Api> {
+        let payment = self.call_value().single_esdt();
+        let user = self.check_and_return_original_owner(&payment);
+        let caller = self.blockchain().get_caller();
+        self.require_user_whitelisted(&user, &caller);
+
+        let claim_result = self.claim_rewards_base_no_farm_token_mint::<FarmStakingWrapper<Self>>(
+            user.clone(),
+            ManagedVec::from_single_item(payment),
+        );
+
+        let mut virtual_farm_token = claim_result.new_farm_token.clone();
+
+        self.set_farm_supply_for_current_week(&claim_result.storage_cache.farm_token_supply);
+
+        self.update_energy_and_progress(&user);
+
+        let new_farm_token_nonce = self.send().esdt_nft_create_compact(
+            &virtual_farm_token.payment.token_identifier,
+            &virtual_farm_token.payment.amount,
+            &virtual_farm_token.attributes,
+        );
+        virtual_farm_token.payment.token_nonce = new_farm_token_nonce;
+
+        let caller = self.blockchain().get_caller();
+        self.send_payment_non_zero(&caller, &virtual_farm_token.payment);
+        self.send_payment_non_zero(&user, &claim_result.rewards);
+
+        self.emit_claim_rewards_event(
+            &caller,
+            claim_result.context,
+            virtual_farm_token.clone(),
+            claim_result.rewards.clone(),
+            claim_result.created_with_merge,
+            claim_result.storage_cache,
+        );
+
+        (virtual_farm_token.payment, claim_result.rewards).into()
+    }
+
+    fn check_and_return_original_owner(&self, payment: &EsdtTokenPayment) -> ManagedAddress {
```

Comment:
Duplicated code with farm, farm-staking-proxy, etc.. Can you refactor to avoid code duplication?

Reply:
Updated for farms and farm staking. Farm staking proxy is too custom to be handled the same way.

Reply:
Also, added a new PermissionsHubModule, to remove duplicated code.